### ISSUE TITLE
Agda generates a `with-clause` that is not well-formed

### ISSUE THREAD

**Issue by ahmadsalim** [agda/agda#2383]
I apologize in advance for the size of the example, but I was unsure how to minimize it further since it contains a lot of type-level manipulation.

In the following file:
https://gist.github.com/ahmadsalim/e4b5cf97317bedf4bfdeae5ce3a088b6

I get the following error when using the code that is on lines 144-145 instead of the one on 161 (basically I introduce a `with-clause` to pattern match on).

The `with`-clause generation then gives me the following error:
```agda
dom f !=
      FSets._◀_ {α} {A} {{decEqA}} (dom f) a
      | Any.any (decEqA DecEq.≟ a) (FSet.elements (dom f))
 of type FSet A
 when checking that the type
      {α β : Level.Level} {A : Set α} {{decEqA : DecEq A}} {B : Set β}
      (f : A ⇀ B) (a : A) (a′ : FSet.Element (dom f))
      (w : Dec (FSet.Element.value a′ ≡ a)) (b b′ : B)
      (a′≢a : FSet.Element.value a′ ≡ a → ⊥) (f·a′≡b′ : f · a′ ≡ b′) →
      (_·′_ {α} {β} {A} {{decEqA}} {B} f a b a′ | w) ≡ b′
of the generated with function is well-formed
```

Basically the issue in particular is that it uses `a′` (which has type `FSet.Element (dom f)`) instead of `FSets.coe (update-dom₂ f a b) a′` (which has the correct typed `FSet.Element (dom f   FSets.◀ a)`) in the type of the generated with-clause, resulting in a type mismatch.

With some help on IRC, I have tried to debug and comment in the file.
I would hope that I have provided enough information for the bug to be found, and I am happy to answer questions relating to the Gist I have provided.

**Comment by UlfNorell** [comment]
Duplicate of #2732.

The problem is that `FSets.coe (update-dom₂ f a b) a′` reduces to `FSet.‹ FSet.Element.value a′ ›`, which is incorrectly eta-contracted to `a′` causing the error.
